driver/src/price_estimation/threaded_price_source.rs

```diff
@@ -0,0 +1,174 @@
+use super::{price_source::PriceSource, Token};
+use crate::models::TokenId;
+use anyhow::Result;
+use std::collections::HashMap;
+#[cfg(test)]
+use std::sync::mpsc::Receiver;
+use std::sync::{
+    mpsc::{self, RecvTimeoutError, Sender},
+    Arc, Mutex,
+};
+use std::thread::{self, JoinHandle};
+use std::time::Duration;
+
+/// Implements `PriceSource` in a non blocking way by updating prices in a
+/// thread and reusing previous results.
+pub struct ThreadedPriceSource {
+    // Shared between this struct and the thread. The background thread writes
+    // the prices, the thread calling `get_prices` reads them.
+    price_map: Arc<Mutex<HashMap<TokenId, u128>>>,
+    // Allows the thread to notice when the owning struct is dropped.
+    _channel_to_thread: Sender<()>,
+
+    // To make testing easier we use another channel to which the thread sends
+    // a message whenever it completes one update loop.
+    #[cfg(test)]
```

Comment:
We could even add this in any case as it doesn't seem that perfomance critical.

Reply:
I find it clearer if it's only there for testing. It actually would be a problem too in non testing because the channel would buffer the update messages forever as no one is reading them.